sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -96,6 +96,13 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
     case _ => None
   }
 
+  protected def variantExprInPartitionExpression(plan: LogicalPlan): Option[Expression] =
```

Comment:
qq, Map and Variant should be excluded in the same scenarios. Can you check if we are covering everything for Variant by checking what do for Map?

Reply:
Repartition by map is currently allowed. I don't know whether that is intended. [My previous PR](https://github.com/apache/spark/pull/48909) which blocked set operations on Variant followed the same pattern as Maps.

Reply:
We should also add a check for Map I suppose.

Reply:
That could be quite risky considering several Spark workloads could currently be repartitioning by map types. It's fine for Variant as Variant is a new feature.
Also, broken "repartitioning" isn't a correctness problem since query results should be agnostic of the partitioning right? I think it would result in suboptimal performance at worst.

Reply:
@harshmotw-db no it will be an actual correctness issue. The problem - like with variant - is that the order of key-value pairs in a map is undefined, e.g. `Map(a ->b, b -> c) == Map(b -> c, a -> b)`. However this is not your problem. Your point about it being a problem for existing workloads is correct, however that would only mean we would have to add a feature flag. cc @cloud-fan @HyukjinKwon

Reply:
Here is the ticket: https://issues.apache.org/jira/browse/SPARK-50525

Reply:
Ah good catch! I think we can fix it now by adding `MapSort` expression for the `Repartition` node in `InsertMapSortInGroupingExpressions` (and rename the rule)

cc @stevomitric @nebojsa-db